PurchasesCoreSwift/Caching/DeviceCache.swift

```diff
@@ -0,0 +1,491 @@
+        }
+    }
+    @objc open var cachedOfferings: Offerings? { offeringsCachedObject.cachedInstance() }
+
+    private let cacheDurationInSecondsInForeground = 60 * 5.0
+    private let cacheDurationInSecondsInBackground = 60 * 60 * 25.0
+    private let accessQueue = DispatchQueue(label: "DeviceCacheQueue")
+    private var threadUnsafeCachedAppUserID: String? { userDefaults.string(forKey: CacheKeys.appUserDefaults.rawValue) }
+    private let userDefaults: UserDefaults
+    private let notificationCenter: NotificationCenter
+    private let offeringsCachedObject: InMemoryCachedObject<Offerings>
+    private var appUserIDHasBeenSet: Bool = false
+    private let assertionFunction: (String) -> Void
+
+    @objc convenience public init(userDefaults: UserDefaults = UserDefaults.standard) {
+        self.init(userDefaults: userDefaults, offeringsCachedObject: nil, notificationCenter: nil)
+    }
+
+    public init(userDefaults: UserDefaults = UserDefaults.standard,
+                offeringsCachedObject: InMemoryCachedObject<Offerings>? = InMemoryCachedObject(),
+                notificationCenter: NotificationCenter? = NotificationCenter.default,
+                assertionFunction: @escaping (String) -> Void = { assertionFailure($0) }) {
+
+        self.offeringsCachedObject = offeringsCachedObject ?? InMemoryCachedObject()
+        self.notificationCenter = notificationCenter ?? NotificationCenter.default
+        self.userDefaults = userDefaults
+        self.assertionFunction = assertionFunction
+        self.appUserIDHasBeenSet = userDefaults.string(forKey: CacheKeys.appUserDefaults) != nil
+
+        super.init()
+
+        self.notificationCenter.addObserver(self,
+                                            selector: #selector(handleUserDefaultsChanged),
+                                            name: UserDefaults.didChangeNotification,
+                                            object: self.userDefaults)
+    }
+
+    @objc private func handleUserDefaultsChanged(notification: Notification) {
+        guard let notificationObject = notification.object as? UserDefaults else {
+            return
+        }
+
+        if appUserIDHasBeenSet && notificationObject == self.userDefaults {
+            if threadUnsafeCachedAppUserID == nil {
+                assertionFunction(
+                    """
+                    [Purchases] - Cached appUserID has been deleted from user defaults.
+                    This leaves the SDK in an undetermined state. Please make sure that RevenueCat
+                    entries in user defaults don't get deleted by anything other than the SDK.
+                    More info: https://rev.cat/userdefaults-crash
+                    """
+                )
+            }
+        }
+    }
+
+    deinit {
+        notificationCenter.removeObserver(self)
+    }
+
+    // MARK: - appUserID
+    @objc(cacheAppUserID:) open func cache(appUserID: String) {
```

Comment:
these don't need to be `open` - when overriding for testing, we use `@testable import`

Reply:
🤦🤦🤦🤦🤦🤦🤦🤦🤦🤦🤦🤦🤦🤦🤦🤦🤦🤦🤦🤦🤦🤦
HOW DO I KEEP FORGETTING `@testable` ?